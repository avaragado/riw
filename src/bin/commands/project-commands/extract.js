// @flow

import path from 'path';

import type yargs from 'yargs';

import outdent from 'outdent';
import ora from 'ora';
import chalk from 'chalk';

import { createHandlerWithRIW, createBar } from '../../utils';

export const command = 'extract';
export const desc = 'Extract react-intl message descriptors from your project components';

const here = `project ${command}`;

export const builder = (yyargs: yargs.Argv) => yyargs
    .usage(outdent`
        $0 ${here}

        Extracts all react-intl message descriptors from your project components.
        This uses babel-plugin-react-intl.

        !!!::TODO:: more info here, including use of config options

        also: uses the same babel options that would apply in your project,
        via .babelrc or babel settings in package.json. if you build using webpack (eg)
        that lets you plug in additional config, it won't spot this.
    `);

const frelFromFabsFromConfig = (config: RIWConfig) => (fabs: AbsolutePath) =>
    chalk.green(path.relative(config.rootDir || '.', fabs));

export const handler = createHandlerWithRIW((riw: RIW) => {
    const frelFromFabs = frelFromFabsFromConfig(riw.config);

    const spinner = ora();
    let ctFile;
    let numFile = 0;
    let ctMD = 0;
    let bar;

    const opt: RIWCLIOptProjectExtract = {
        on: {
            start: () => {
                spinner.text = 'Finding source files...';
                spinner.start();
            },
            startFiles: (arfabs) => {
                ctFile = arfabs.length;
                bar = createBar(ctFile);
                spinner.text = `Processing ${ctFile} files...`;
            },
            startFile: (fabs) => {
                numFile += 1;
                spinner.text = `${bar(numFile)} [${ctMD}] ${frelFromFabs(fabs)}`;
            },
            endFile: (smdd: RIWSourceMessageDescriptorData) => {
                ctMD += smdd.armd.length;
                spinner.text = `${bar(numFile)} [${ctMD}] ${frelFromFabs(smdd.fabs)}`;
                spinner.render();
            },
            startDupCheck: () => {
                spinner.text = `${bar(numFile)} [${ctMD}] Checking for duplicate message descriptor ids...`;
                spinner.render();
            },
        },
    };

    const { armd, dups } = riw.project.extract(opt);

    spinner.succeed(`Found ${chalk.bold(ctMD.toString(), 'message descriptors')} from ${chalk.bold(numFile.toString(), 'files')}`);

    if (dups.length) {
        spinner.fail(`Duplicate message descriptor ids: ${dups.length}`);
        console.log(dups.map(
            dup => outdent`
              ${outdent}
                - ${chalk.blue(dup.id)} used in:
                  ${dup.arfabs.map(frelFromFabs).join('\n    ')}
            `,
        ).join('\n'));
    } else {
        spinner.succeed('No duplicate message descriptor ids');
    }

    // console.log('FINAL DATA:', armd, dups);
});
