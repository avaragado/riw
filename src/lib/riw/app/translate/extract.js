// @flow

import chain from 'ramda/src/chain';
import compose from 'ramda/src/compose';
import map from 'ramda/src/map';
import prop from 'ramda/src/prop';

import { arfabsInputJSON, arfabsInputSource } from '../../../config-helper';

import armdFromJSONFabs from './armdFromJSONFabs';
import armdfromSourceFabs from './armdFromSourceFabs';

type MDExtractor = (arfabsFromConfig: FilesFromConfig, armdFromFabs: MessageDescriptorsFromFile) =>
    <T>(notify: string => T => T) =>
    (config: RIWConfig) => RIWMessageDescriptor[];

const armdExtract: MDExtractor = (arfabsFromConfig, armdFromFabs) => notify => compose(
    notify('endExtract'),
    chain(prop('armd')),
    map(compose(
        notify('endExtractFile'),
        fabs => ({
            fabs,
            armd: map(
                md => ({ ...md, file: fabs }),
                armdFromFabs(fabs),
            ),
        }),
        notify('startExtractFile'),
    )),
    notify('startExtract'),
    arfabsFromConfig,
);

export const armdExtractSource = armdExtract(arfabsInputSource, armdfromSourceFabs);
export const armdExtractJSON = armdExtract(arfabsInputJSON, armdFromJSONFabs);
