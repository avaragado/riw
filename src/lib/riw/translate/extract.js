// @flow

import chain from 'ramda/src/chain';
import compose from 'ramda/src/compose';
import map from 'ramda/src/map';
import prop from 'ramda/src/prop';

import * as source from './extractFromSource';
import * as json from './extractFromJSON';

type MDExtractor = (arfabsFromConfig: FilesFromConfig, armdFromFabs: MessageDescriptorsFromFile) =>
    <T>(notify: string => T => T) =>
    (config: RIWConfig) => RIWMessageDescriptor[];

const armdExtract: MDExtractor = (arfabsFromConfig, armdFromFabs) => notify => compose(
    notify('endExtract'),
    chain(prop('armd')),
    map(compose(
        notify('endExtractFile'),
        fabs => ({
            fabs,
            armd: map(
                md => ({ ...md, fabs }),
                armdFromFabs(fabs),
            ),
        }),
        notify('startExtractFile'),
    )),
    notify('startExtract'),
    arfabsFromConfig,
);

export const armdExtractSource = armdExtract(source.arfabsFromConfig, source.armdFromFabs);
export const armdExtractJSON = armdExtract(json.arfabsFromConfig, json.armdFromFabs);
