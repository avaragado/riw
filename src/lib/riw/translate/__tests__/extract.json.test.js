// @flow

import path from 'path';

import mock from 'mock-fs';
import outdent from 'outdent';

import { armdExtractJSON } from '../extract';
import cfgBase from '../../__tests__/helpers/dummyConfig';

const absify = (name, frel) => path.resolve('fixtures', name, frel);

const cfgOverride = {
    ...cfgBase,
    rootDir: '.',
};

const notify = () => x => x;

type Fixture = {
    name: string,
    dir: { [key: string]: string },
    configOverride?: Object,
    out: RIWMessageDescriptor[],
};

const fixtures: Fixture[] = [
    {
        name: '01',
        dir: {}, // empty
        out: [],
    },

    {
        name: '02',
        dir: {
            'a.json': '[]',
            'b.json': '[]',
        },
        out: [],
    },

    {
        name: '03',
        dir: {
            'a.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1!"
                    }
                ]
            `,
            'aa.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1 again!"
                    }
                ]
            `,
            'b.json': outdent`
                [
                    {
                        "id": "b.1",
                        "defaultMessage": "b.1!"
                    },
                    {
                        "id": "b.2",
                        "description": "b.2 desc",
                        "defaultMessage": "b.2!"
                    }
                ]
            `,
            'bb.json': outdent`
                [
                    {
                        "id": "b.2",
                        "description": "b.2 again desc",
                        "defaultMessage": "b.2 again!"
                    }
                ]
            `,
            'c.json': outdent`
                [
                    {
                        "id": "c.1",
                        "defaultMessage": "c.1!"
                    }
                ]
            `,
        },
        configOverride: {
            collateDir: 'fixtures/03',
            inputMode: 'json',
        },
        out: [
            {
                id: 'a.1',
                description: undefined,
                defaultMessage: 'a.1!',
                fabs: absify('03', 'a.json'),
            },
            {
                id: 'a.1',
                description: undefined,
                defaultMessage: 'a.1 again!',
                fabs: absify('03', 'aa.json'),
            },
            {
                id: 'b.1',
                description: undefined,
                defaultMessage: 'b.1!',
                fabs: absify('03', 'b.json'),
            },
            {
                id: 'b.2',
                description: 'b.2 desc',
                defaultMessage: 'b.2!',
                fabs: absify('03', 'b.json'),
            },
            {
                id: 'b.2',
                description: 'b.2 again desc',
                defaultMessage: 'b.2 again!',
                fabs: absify('03', 'bb.json'),
            },
            {
                id: 'c.1',
                description: undefined,
                defaultMessage: 'c.1!',
                fabs: absify('03', 'c.json'),
            },
        ],
    },

    {
        name: '04',
        dir: {
            'a.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1!"
                    }
                ]
            `,
            'aa.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1 again!"
                    }
                ]
            `,
            'b.json': outdent`
                [
                    {
                        "id": "b.1",
                        "defaultMessage": "b.1!"
                    },
                    {
                        "id": "b.2",
                        "description": "b.2 desc",
                        "defaultMessage": "b.2!"
                    }
                ]
            `,
            'bb.json': outdent`
                [
                    {
                        "id": "b.2",
                        "description": "b.2 again desc",
                        "defaultMessage": "b.2 again!"
                    }
                ]
            `,
            'c.json': outdent`
                [
                    {
                        "id": "c.1",
                        "defaultMessage": "c.1!"
                    }
                ]
            `,
        },
        configOverride: {
            collateDir: absify('04', ''),
            inputMode: 'json',
        },
        out: [
            {
                id: 'a.1',
                description: undefined,
                defaultMessage: 'a.1!',
                fabs: absify('04', 'a.json'),
            },
            {
                id: 'a.1',
                description: undefined,
                defaultMessage: 'a.1 again!',
                fabs: absify('04', 'aa.json'),
            },
            {
                id: 'b.1',
                description: undefined,
                defaultMessage: 'b.1!',
                fabs: absify('04', 'b.json'),
            },
            {
                id: 'b.2',
                description: 'b.2 desc',
                defaultMessage: 'b.2!',
                fabs: absify('04', 'b.json'),
            },
            {
                id: 'b.2',
                description: 'b.2 again desc',
                defaultMessage: 'b.2 again!',
                fabs: absify('04', 'bb.json'),
            },
            {
                id: 'c.1',
                description: undefined,
                defaultMessage: 'c.1!',
                fabs: absify('04', 'c.json'),
            },
        ],
    },
];

describe('lib/riw/translate/extract.json', () => {
    afterEach(() => {
        mock.restore();
    });

    fixtures.forEach((fixture) => {
        it(fixture.name, () => {
            mock({
                fixtures: {
                    [fixture.name]: fixture.dir,
                },
            });

            const cfg = {
                ...cfgOverride,
                sourceDirs: [`fixtures/${fixture.name}/**/*.js`],
                ...fixture.configOverride,
            };

            const received = armdExtractJSON(notify)(cfg);

            expect(received).toEqual(fixture.out);
        });
    });
});
