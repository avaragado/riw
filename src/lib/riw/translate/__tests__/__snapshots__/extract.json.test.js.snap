// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/riw/translate/extract.json 01 1`] = `Array []`;

exports[`lib/riw/translate/extract.json 02 1`] = `Array []`;

exports[`lib/riw/translate/extract.json 03 1`] = `
Array [
  Object {
    "defaultMessage": "a.1!",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/a.json",
    "id": "a.1",
  },
  Object {
    "defaultMessage": "a.1 again!",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.json",
    "id": "a.1",
  },
  Object {
    "defaultMessage": "b.1!",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.json",
    "id": "b.1",
  },
  Object {
    "defaultMessage": "b.2!",
    "description": "b.2 desc",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.json",
    "id": "b.2",
  },
  Object {
    "defaultMessage": "b.2 again!",
    "description": "b.2 again desc",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.json",
    "id": "b.2",
  },
  Object {
    "defaultMessage": "c.1!",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.json",
    "id": "c.1",
  },
]
`;

exports[`lib/riw/translate/extract.json 04 1`] = `
Array [
  Object {
    "defaultMessage": "a.1!",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/a.json",
    "id": "a.1",
  },
  Object {
    "defaultMessage": "a.1 again!",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.json",
    "id": "a.1",
  },
  Object {
    "defaultMessage": "b.1!",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.json",
    "id": "b.1",
  },
  Object {
    "defaultMessage": "b.2!",
    "description": "b.2 desc",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.json",
    "id": "b.2",
  },
  Object {
    "defaultMessage": "b.2 again!",
    "description": "b.2 again desc",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.json",
    "id": "b.2",
  },
  Object {
    "defaultMessage": "c.1!",
    "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.json",
    "id": "c.1",
  },
]
`;
